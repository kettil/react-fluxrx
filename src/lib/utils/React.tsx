import { FunctionComponent, memo } from 'react';

export const hocOptimize = <C, W>(Component: FunctionComponent<C>, Wrapper: FunctionComponent<W>) => {
  const displayName = Component.displayName || Component.name;

  // istanbul ignore else
  if (displayName) {
    Wrapper.displayName = displayName;
  }

  return memo(Wrapper);
};
