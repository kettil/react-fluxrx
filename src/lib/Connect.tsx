import hoistNonReactStatics from 'hoist-non-react-statics';
import React, { memo } from 'react';
import useConnect from './hooks/useConnect';
import { ExtractProps, MapDispatchToPropsType, MapStateToPropsType, StoreType } from './types';
import { defaultMapDispatchToProps, defaultMapStateToProps } from './utils/connect';

export const connectWrapper = <State,>(context: React.Context<StoreType<State>>) => {
  return <OProps extends Record<any, any>, MState = {}, MDispatch = {}>(
    mapStateToProps: MapStateToPropsType<State, OProps, MState> = defaultMapStateToProps,
    mapDispatchToProps: MapDispatchToPropsType<OProps, MDispatch> = defaultMapDispatchToProps,
  ) => {
    return <Component extends React.ComponentType<IProps>, IProps = ExtractProps<Component>>(Component: Component) => {
      return connect<Component, State, IProps, OProps, MState, MDispatch>(
        context,
        mapStateToProps,
        mapDispatchToProps,
        Component,
      );
    };
  };
};

export const connect = <Component extends React.ComponentType<IProps>, State, IProps, OProps, MState, MDispatch>(
  context: React.Context<StoreType<State>>,
  mapStateToProps: MapStateToPropsType<State, OProps, MState>,
  mapDispatchToProps: MapDispatchToPropsType<OProps, MDispatch>,
  Component: Component,
) => {
  const Consumer: React.FC<OProps> = (outerProps) => {
    const innerProps: IProps = useConnect(context, mapStateToProps, mapDispatchToProps, outerProps);

    return <Component {...(innerProps as any)} />;
  };

  const displayName = Component.displayName || Component.name;

  // istanbul ignore else
  if (displayName) {
    Consumer.displayName = displayName;
  }

  return hoistNonReactStatics(memo(Consumer), Component);
};

export default connectWrapper;
