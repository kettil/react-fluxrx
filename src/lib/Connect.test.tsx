/* tslint:disable:no-implicit-dependencies no-unused-expression */
import React from 'react';
import renderer from 'react-test-renderer';

import { BehaviorSubject } from 'rxjs';

import { createConnect } from './Connect';

/**
 *
 */
describe('Check the connect function', () => {
  /**
   *
   */
  test('it should be return a connected component when createConnect() and his callbacks is called', () => {
    const dispatch = jest.fn();
    const subject = new BehaviorSubject({ isState: true });
    const context: React.Context<any> = React.createContext({
      dispatch,
      subscribe: (cb: any) => subject.subscribe(cb),
      getState: () => subject.getValue(),
    });

    const connect = createConnect(context);
    expect(connect).toBeInstanceOf(Function);

    const callback = connect<{ message: string }>();
    expect(callback).toBeInstanceOf(Function);
    expect(callback.length).toBe(1);

    const Component = ({ message }: { message: string }) => <div>Hello {message}</div>;
    const ConnectComponent = callback(Component);

    let root: renderer.ReactTestRenderer;

    renderer.act(() => {
      root = renderer.create(<ConnectComponent message={'World'} />);
    });

    expect(root!.toJSON()).toMatchSnapshot();
  });
});
