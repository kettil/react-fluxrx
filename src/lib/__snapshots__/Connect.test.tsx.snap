// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the connect function it should be return a connected component when connectWrapper() and his callbacks is called 1`] = `
<div>
  Hello World
</div>
`;
